File: temp/155_最小栈.php
<?php

class MinStack
{
    private $stack = [];

    /**
     * 时间复杂度: O(1)
     */
    public function getMin()
    {
        return $this->stack[count($this->stack) - 1][1] ?? null;
    }

    public function push($value)
    {
        $min = $this->getMin();
        $this->stack[] = [$value, $min !== null ? min($value, $min) : $value];
    }

    public function top()
    {
        return $this->stack[count($this->stack) - 1][0];
    }

    public function pop()
    {
        $pop = array_pop($this->stack);
        return $pop[0];
    }
}

$stack = new MinStack();

$stack->push(-2);
$stack->push(0);
$stack->push(-3);
var_dump($stack->getMin()); // -3
$stack->pop();
var_dump($stack->top()); // 0
var_dump($stack->getMin()); // -2

File: leetcode-golang/stack/155_最小栈.go
package stack

type MinStack struct {
	Nodes []Node
}

type Node struct {
	Val int
	Min int
}


/** initialize your data structure here. */
func Constructor() MinStack {
	return MinStack{
		Nodes: []Node{},
	}
}


func (this *MinStack) Push(x int)  {
	var node Node
	node.Val = x
	if len(this.Nodes) == 0 {
		node.Min = x
	} else {
		node.Min = int(math.Min(float64(this.Nodes[len(this.Nodes) - 1].Min), float64(x)))
	}
	this.Nodes = append(this.Nodes, node)
}


func (this *MinStack) Pop()  {
	if len(this.Nodes) > 0 {
		this.Nodes = this.Nodes[:len(this.Nodes)-1]
	}
}


func (this *MinStack) Top() int {
	if len(this.Nodes) == 0 {
		return 0
	}
	return this.Nodes[len(this.Nodes)-1].Val
}


func (this *MinStack) GetMin() int {
	if len(this.Nodes) == 0 {
		return 0
	}
	return this.Nodes[len(this.Nodes)-1].Min
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */